src/core/lib/channel/channel_tracer.cc

```diff
@@ -0,0 +1,327 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/channel/channel_tracer.h"
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/useful.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "src/core/lib/iomgr/error.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/support/object_registry.h"
+#include "src/core/lib/support/string.h"
+#include "src/core/lib/surface/channel.h"
+#include "src/core/lib/transport/connectivity_state.h"
+
+grpc_core::DebugOnlyTraceFlag grpc_trace_channel_tracer_refcount(
+    false, "channel_tracer_refcount");
+
+// One node of tracing data
+typedef struct grpc_trace_node {
+  grpc_slice data;
+  grpc_error* error;
+  gpr_timespec time_created;
+  grpc_connectivity_state connectivity_state;
+  struct grpc_trace_node* next;
+
+  // the tracer object for the (sub)channel that this trace node refers to.
+  grpc_channel_tracer* referenced_tracer;
+} grpc_trace_node;
+
+/* the channel tracing object */
+struct grpc_channel_tracer {
```

Comment:
would it make sense to convert this into a class, and bring the related functions inside it? Bring it under the grpc_core namespace maybe.